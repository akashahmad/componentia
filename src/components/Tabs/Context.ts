import * as React from "react";

const initialValues: ValueTypes = {
    titles: [],
    activeIndex: 0,
    setTitles: () => {},
    setActiveIndex: () => {},
};

export const TabContext = React.createContext(initialValues);

interface ValueTypes {
    titles: string[];
    activeIndex: number;
    setTitles: React.Dispatch<React.SetStateAction<string[]>>;
    setActiveIndex: React.Dispatch<React.SetStateAction<number>>;
}
