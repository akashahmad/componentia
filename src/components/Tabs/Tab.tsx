import * as React from "react";

import { TabContext } from "./Context";

export function Tab({ title, children }: TabProps) {
    const context = React.useContext(TabContext);
    if (!context) throw new Error("Tab must be used inside the Tabs component.");

    React.useEffect(() => {
        if (!context.titles.includes(title)) {
            context.setTitles(prev => [...prev, title]);
        }
    }, [title]);

    const currentIndex = context.titles.findIndex(x => x === title);

    return currentIndex === context.activeIndex && <div>{children}</div>;
}

interface TabProps {
    title: string;
    children: React.ReactNode;
}
