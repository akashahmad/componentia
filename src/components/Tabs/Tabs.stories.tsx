import * as React from "react";

import type { Meta, StoryObj } from "@storybook/react";

import { Tabs } from "../Tabs/Tabs";

const meta: Meta<typeof Tabs> = {
    component: Tabs,
    tags: ["autodocs"],
};

export default meta;
type Story = StoryObj<typeof Tabs>;

export const Default: Story = {
    args: {},
    argTypes: {
        children: {
            control: { type: undefined },
            description: "``Tabs.Tab``",
            table: {
                type: { summary: "string" },
                defaultValue: { summary: "undefined" },
            },
        },
        titleRenderer: {
            control: false,
            table: {
                type: { summary: "string" },
                defaultValue: { summary: "DefaultRenderer" },
            },
        },
    },
    render: props => (
        <Tabs {...props}>
            <Tabs.Tab title="Profile">Tab One Content</Tabs.Tab>
            <Tabs.Tab title="Password">Tab Two Content</Tabs.Tab>
            <Tabs.Tab title="Team">Tab Three Content</Tabs.Tab>
            <Tabs.Tab title="Notification">Tab Four Content</Tabs.Tab>
            <Tabs.Tab title="Integrations">Tab Five Content</Tabs.Tab>
            <Tabs.Tab title="Licences">Tab Six Content</Tabs.Tab>
        </Tabs>
    ),
};
