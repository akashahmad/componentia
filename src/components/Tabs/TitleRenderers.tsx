import * as React from "react";

import cx from "classnames";

import { RendererProps } from "./Tabs";

export const DefaultRenderer = ({ titles, activeIndex, setActiveIndex }: RendererProps) => {
    return (
        <div className="flex items-center bg-white">
            {titles.map((title, index) => {
                const isActive = activeIndex === index;

                return (
                    <p
                        key={title + index}
                        className={cx(
                            "inline-block first:ml-[25px] text-sm leading-6 font-medium cursor-pointer py-[9px] mx-5",
                            {
                                ["text-primary-text"]: !isActive,
                                [`border-b-2 text-primary border-primary`]: isActive,
                                ["border-b-2 border-white"]: !isActive,
                            },
                        )}
                        onClick={() => setActiveIndex(index)}
                    >
                        {title}
                    </p>
                );
            })}
        </div>
    );
};
