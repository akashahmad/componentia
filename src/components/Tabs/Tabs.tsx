import * as React from "react";

import { Tab } from "./Tab";
import { DefaultRenderer } from "./TitleRenderers";
import { TabContext } from "./Context";

export function Tabs({ children, titleRenderer: TitleRenderer = DefaultRenderer, className }: TabsProps) {
    const [titles, setTitles] = React.useState<string[]>([]);
    const [activeIndex, setActiveIndex] = React.useState<number>(0);

    return (
        <TabContext.Provider value={{ titles, setTitles, activeIndex, setActiveIndex }}>
            <div className={className}>
                <TitleRenderer titles={titles} activeIndex={activeIndex} setActiveIndex={setActiveIndex} />
                <div>{children}</div>
            </div>
        </TabContext.Provider>
    );
}

interface TabsProps {
    children: React.ReactNode;
    titleRenderer?: React.FC<RendererProps>;
    className?: string;
}

export interface RendererProps {
    titles: string[];
    activeIndex: number;
    setActiveIndex: React.Dispatch<React.SetStateAction<number>>;
}

Tabs.Tab = Tab;
export { Tab };
