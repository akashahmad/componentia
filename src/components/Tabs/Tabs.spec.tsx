import * as React from "react";

import { render, screen } from "@testing-library/react";
import { Tabs } from "./Tabs";

test("group-button component is rendering properly", () => {
    const TITLE_1 = "TITLE_1";
    const TITLE_2 = "TITLE_2";

    const { container } = render(
        <Tabs>
            <Tabs.Tab title={TITLE_1}>
                Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the
                industrys standard dummy text ever since the 1500s, when
            </Tabs.Tab>
            <Tabs.Tab title={TITLE_2}>
                Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the
                industrys standard dummy text ever since the 1500s, when
            </Tabs.Tab>
        </Tabs>,
    );

    expect(screen.getByText(TITLE_1)).not.toBeNull();
    expect(screen.getByText(TITLE_2)).not.toBeNull();

    expect(container).toMatchSnapshot();
});
