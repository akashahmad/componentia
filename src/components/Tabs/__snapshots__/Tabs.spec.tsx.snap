// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`group-button component is rendering properly 1`] = `
<div>
  <div>
    <div
      class="flex items-center bg-white"
    >
      <p
        class="inline-block first:ml-[25px] text-sm leading-6 font-medium cursor-pointer py-[9px] mx-5 border-b-2 text-primary border-primary"
      >
        TITLE_1
      </p>
      <p
        class="inline-block first:ml-[25px] text-sm leading-6 font-medium cursor-pointer py-[9px] mx-5 text-primary-text border-b-2 border-white"
      >
        TITLE_2
      </p>
    </div>
    <div>
      <div>
        Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industrys standard dummy text ever since the 1500s, when
      </div>
    </div>
  </div>
</div>
`;
